lib/Runtime/Base/FunctionBody.cpp

```diff
@@ -55,6 +56,213 @@ namespace Js
+        }
+        else
+        {
+            Assert(false);
+        }
+    }
+
+    template<class T, typename FIELDS>
+    void AuxPtrs<T, FIELDS>::allocAuxPtr(T* _this, uint8 count, Recycler* recycler)
+    {
+        Assert(count >= AuxPtrs32::MaxCount);
+        auto requestSize = sizeof(AuxPtrs<T, FIELDS>) + (count - 1)*sizeof(void*);
+        auto allocSize = ::Math::Align<uint8>((uint8)requestSize, 16);
+        auto capacity = (uint8)((allocSize - offsetof(AuxPtrsT, ptrs)) / sizeof(void*));
+
+        if (_this->auxPtrs->count != AuxPtrs32::MaxCount) // expanding
+        {
+            _this->auxPtrs = RecyclerNewWithBarrierPlusZ(recycler, allocSize - sizeof(AuxPtrsT), AuxPtrsT, capacity, _this->auxPtrs);
+        }
+        else // promoting
+        {
+            _this->auxPtrs = RecyclerNewWithBarrierPlusZ(recycler, allocSize - sizeof(AuxPtrsT), AuxPtrsT, capacity, (AuxPtrs32*)(void*)_this->auxPtrs);
+        }
+    }
+
+    //static void* get_AuxPtr(T* _this, FIELDS e, Recycler* recycler) const;
+    //static void set_AuxPtr(T* _this, FIELDS e, void* ptr, Recycler* recycler);
+    template<class T, typename FIELDS>
+    inline void* AuxPtrs<T, FIELDS>::get_AuxPtr(const T* _this, FIELDS e)
+    {
+        if (_this->auxPtrs == nullptr)
+        {
+            return nullptr;
+        }
+        if (_this->auxPtrs->count == AuxPtrs16::MaxCount)
+        {
+            return ((AuxPtrs16*)(void*)_this->auxPtrs)->get(e);
+        }
+        if (_this->auxPtrs->count == AuxPtrs32::MaxCount)
+        {
+            return ((AuxPtrs32*)(void*)_this->auxPtrs)->get(e);
+        }
+        return _this->auxPtrs->get(e);
+    }
+    template<class T, typename FIELDS>
+    void AuxPtrs<T, FIELDS>::set_AuxPtr(T* _this, FIELDS e, void* ptr, Recycler* recycler)
+    {
+        if (ptr == nullptr && get_AuxPtr(_this, e) == nullptr)
+        {
+            return;
+        }
+        if (_this->auxPtrs == nullptr)
+        {
+            AuxPtrs<FunctionProxy, FIELDS>::allocAuxPtrFix(_this, 16, recycler);
+            bool ret = ((AuxPtrs16*)(void*)_this->auxPtrs)->set(e, ptr);
+            Assert(ret);
+            return;
+        }
+        if (_this->auxPtrs->count == AuxPtrs16::MaxCount)
+        {
+            bool ret = ((AuxPtrs16*)(void*)_this->auxPtrs)->set(e, ptr);
+            if (ret)
+            {
+                return;
+            }
+            else
+            {
+                AuxPtrs<FunctionProxy, FIELDS>::allocAuxPtrFix(_this, 32, recycler);
+            }
+        }
+        if (_this->auxPtrs->count == AuxPtrs32::MaxCount)
+        {
+            bool ret = ((AuxPtrs32*)(void*)_this->auxPtrs)->set(e, ptr);
+            if (ret)
+            {
+                return;
+            }
+            else
+            {
+                AuxPtrs<FunctionProxy, FIELDS>::allocAuxPtr(_this, AuxPtrs32::MaxCount + 1, recycler);
```

Comment:
if more than 3 pointers need to be hold in the AuxPointers, it will promote to the flexible one -- AuxPtrs, which use bigger metadata (24 bytes metadata which holds up to 22 pointers) but better performance (no linear search). currently the hard upper limit is 253 pointers (the metadata will grow every 8 pointers though)